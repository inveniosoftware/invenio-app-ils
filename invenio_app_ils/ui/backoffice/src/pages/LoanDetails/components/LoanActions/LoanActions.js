import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { List, Button } from 'semantic-ui-react';

export default class LoanActions extends Component {
  constructor(props) {
    super(props);
    this.performLoanAction = this.props.performLoanAction;
  }

  renderAvailableActions(pid, loan, actions = {}) {
    return Object.keys(actions).map(action => {
      return (
        <List.Item key={action}>
          <Button
            primary
            onClick={() => {
              this.performLoanAction(pid, loan, actions[action]);
            }}
          >
            {action}
          </Button>
        </List.Item>
      );
    });
  }

  render() {
    const {
      availableActions: actions,
      metadata: loan,
      id: pid,
    } = this.props.loanDetails;
    return (
      <List horizontal>{this.renderAvailableActions(pid, loan, actions)}</List>
    );
  }
}

LoanActions.propTypes = {
  loanDetails: PropTypes.object.isRequired,
  performLoanAction: PropTypes.func,
};
