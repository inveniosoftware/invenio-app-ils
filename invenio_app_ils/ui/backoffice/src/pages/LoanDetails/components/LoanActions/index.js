export { default as LoanActions } from './LoanActions';
