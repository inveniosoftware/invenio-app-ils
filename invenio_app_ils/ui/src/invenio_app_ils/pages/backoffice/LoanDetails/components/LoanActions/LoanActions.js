import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { List, Button } from 'semantic-ui-react';
import { omit } from 'lodash/object';
import { CancelLoanModal } from '../CancelLoanModal';

export default class LoanActions extends Component {
  constructor(props) {
    super(props);
    this.performLoanAction = this.props.performLoanAction;
  }

  renderAvailableActions(pid, loan, actions = {}) {
    if ('checkout' in actions && loan.metadata.state === 'PENDING') {
      actions = omit(actions, 'checkout');
    }
    return Object.keys(actions).map(action => {
      const cancelAction = (cancelReason = null) =>
        this.performLoanAction(pid, loan, actions[action], cancelReason);
      const loanAction = () =>
        this.performLoanAction(pid, loan, actions[action]);
      return (
        <List.Item key={action}>
          {action === 'cancel' ? (
            <CancelLoanModal
              loan={this.props.loanDetails}
              action={cancelAction}
            />
          ) : (
            <Button primary onClick={loanAction}>
              {action}
            </Button>
          )}
        </List.Item>
      );
    });
  }

  render() {
    const { availableActions, pid } = this.props.loanDetails;
    const {
      document_pid,
      patron_pid,
      item_pid,
      state,
    } = this.props.loanDetails.metadata;
    const loan = {
      metadata: {
        document_pid: document_pid,
        patron_pid: patron_pid,
        item_pid: item_pid,
        state: state,
      },
    };
    if (availableActions) {
      return (
        <List horizontal>
          {Object.keys(availableActions).length ? (
            this.renderAvailableActions(pid, loan, availableActions)
          ) : (
            <List.Header as="h3">No actions available</List.Header>
          )}
        </List>
      );
    } else {
      return (
        <List horizontal>
          <List.Header as="h3">No actions available</List.Header>
        </List>
      );
    }
  }
}

LoanActions.propTypes = {
  loanDetails: PropTypes.object.isRequired,
  performLoanAction: PropTypes.func,
};
