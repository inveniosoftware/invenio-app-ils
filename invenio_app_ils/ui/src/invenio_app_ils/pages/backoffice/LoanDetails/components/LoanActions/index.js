import { connect } from 'react-redux';
import { performLoanAction } from '../../state/actions';
import LoanActionsComponent from './LoanActions';

const mapStateToProps = state => ({
  error: state.loanDetails.error,
  loanDetails: state.loanDetails.data,
});

const mapDispatchToProps = dispatch => ({
  performLoanAction: (pid, loan, url) =>
    dispatch(performLoanAction(pid, loan, url)),
});

export const LoanActions = connect(
  mapStateToProps,
  mapDispatchToProps
)(LoanActionsComponent);
