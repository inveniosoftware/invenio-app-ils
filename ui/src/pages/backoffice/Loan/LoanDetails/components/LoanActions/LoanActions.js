import {
  InfoMessage,
  OverdueLoanSendMailModal,
} from '@pages/backoffice/components';
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { List, Button, Grid } from 'semantic-ui-react';
import { omit } from 'lodash/object';
import { CancelModal } from '@components/CancelModal';
import _isEmpty from 'lodash/isEmpty';

export default class LoanActions extends Component {
  constructor(props) {
    super(props);
    this.performLoanAction = this.props.performLoanAction;
  }

  renderAvailableActions(pid, patronPid, documentPid, itemPid, actions = {}) {
    // omit checkout because it must done in one of the available items
    if (!itemPid) {
      actions = omit(actions, 'checkout');
    }

    return Object.keys(actions).map(action => {
      const cancelAction = (cancelReason = null) =>
        this.performLoanAction(actions[action], documentPid, patronPid, {
          cancelReason: cancelReason,
        });
      const loanAction = () =>
        this.performLoanAction(actions[action], documentPid, patronPid);
      return (
        <List.Item key={action}>
          {action === 'cancel' ? (
            <CancelModal
              header={`Cancel Loan #${pid}`}
              content={`You are about to cancel loan #${pid}.
                Please enter a reason for cancelling this loan.`}
              cancelText="Cancel Loan"
              buttonText="cancel"
              action={cancelAction}
            />
          ) : (
            <Button primary onClick={loanAction}>
              {action}
            </Button>
          )}
        </List.Item>
      );
    });
  }

  render() {
    const { availableActions, pid } = this.props.loanDetails;
    const {
      document_pid,
      item_pid,
      patron_pid,
    } = this.props.loanDetails.metadata;

    const loanActions = !_isEmpty(availableActions) && (
      <List horizontal>
        {this.renderAvailableActions(
          pid,
          patron_pid,
          document_pid,
          item_pid,
          availableActions
        )}
      </List>
    );
    const sendReminderButton = this.props.loanDetails.metadata.is_overdue && (
      <OverdueLoanSendMailModal
        loan={this.props.loanDetails}
        buttonTriggerText="Send return reminder"
      />
    );
    if (
      !_isEmpty(availableActions) ||
      this.props.loanDetails.metadata.is_overdue
    ) {
      return (
        <Grid columns={2}>
          <Grid.Column width={12}>{loanActions}</Grid.Column>
          <Grid.Column width={4} textAlign="right">
            {sendReminderButton}
          </Grid.Column>
        </Grid>
      );
    } else {
      return (
        <InfoMessage
          fluid
          header={'No actions available.'}
          content={"The loan can't be changed in it's current state."}
        />
      );
    }
  }
}

LoanActions.propTypes = {
  loanDetails: PropTypes.object.isRequired,
  performLoanAction: PropTypes.func,
};
