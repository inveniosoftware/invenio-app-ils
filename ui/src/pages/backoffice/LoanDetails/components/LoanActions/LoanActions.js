import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { List, Button } from 'semantic-ui-react';
import { omit } from 'lodash/object';
import { CancelModal } from '@components/CancelModal';

export default class LoanActions extends Component {
  constructor(props) {
    super(props);
    this.performLoanAction = this.props.performLoanAction;
  }

  renderAvailableActions(pid, patronPid, documentPid, itemPid, actions = {}) {
    // omit checkout because it must done in one of the available items
    if (!itemPid) {
      actions = omit(actions, 'checkout');
    }

    return Object.keys(actions).map(action => {
      const cancelAction = (cancelReason = null) =>
        this.performLoanAction(actions[action], documentPid, patronPid, {
          cancelReason: cancelReason,
        });
      const loanAction = () =>
        this.performLoanAction(actions[action], documentPid, patronPid);
      return (
        <List.Item key={action}>
          {action === 'cancel' ? (
            <CancelModal
              header={`Cancel Loan #${pid}`}
              content={`You are about to cancel loan #${pid}.
                Please enter a reason for cancelling this loan.`}
              cancelText="Cancel Loan"
              buttonText="cancel"
              action={cancelAction}
            />
          ) : (
            <Button primary onClick={loanAction}>
              {action}
            </Button>
          )}
        </List.Item>
      );
    });
  }

  render() {
    const { availableActions, pid } = this.props.loanDetails;
    const {
      document_pid,
      item_pid,
      patron_pid,
    } = this.props.loanDetails.metadata;
    if (availableActions) {
      return (
        <List horizontal>
          {Object.keys(availableActions).length ? (
            this.renderAvailableActions(
              pid,
              patron_pid,
              document_pid,
              item_pid,
              availableActions
            )
          ) : (
            <List.Header as="h3">No actions available</List.Header>
          )}
        </List>
      );
    } else {
      return (
        <List horizontal>
          <List.Header as="h3">No actions available</List.Header>
        </List>
      );
    }
  }
}

LoanActions.propTypes = {
  loanDetails: PropTypes.object.isRequired,
  performLoanAction: PropTypes.func,
};
