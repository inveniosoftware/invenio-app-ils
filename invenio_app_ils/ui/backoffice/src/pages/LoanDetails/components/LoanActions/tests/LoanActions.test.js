import React from 'react';
import { mount } from 'enzyme';
import { List } from 'semantic-ui-react';
import { LoanActions } from '../LoanActions';

const defaultProps = {
  data: {},
  onAction: () => 'onAction',
  actions: {},
};

describe('LoanActions', () => {
  let component;
  afterEach(() => {
    component.unmount();
  });

  it('should render initial state and check props', () => {
    component = mount(<LoanActions {...defaultProps} />);
    Object.keys(defaultProps).forEach(key =>
      expect(component.props()[key]).toEqual(defaultProps[key])
    );
    expect(component).toMatchSnapshot();
  });

  it('should render an List.Item for each available action', () => {
    const props = {
      ...defaultProps,
      data: {
        title: 'A title',
      },
      availableActions: {
        action: 'url',
        anotheraction: 'another-url',
      },
    };

    component = mount(<LoanActions {...props} />);
    expect(component).toMatchSnapshot();
    const actionComponents = component.find(List.Item);
    expect(actionComponents).toHaveLength(Object.keys(props.actions).length);
  });

  it('should render a loader when trigger an action', () => {
    const props = {
      ...defaultProps,
      actionLoading: true,
    };
    component = mount(<LoanActions {...props} />);
    expect(component).toMatchSnapshot();
    const actionComponents = component.find(List.Item);
    // should not render any List.Item component
    expect(actionComponents).toHaveLength(0);
  });
});
