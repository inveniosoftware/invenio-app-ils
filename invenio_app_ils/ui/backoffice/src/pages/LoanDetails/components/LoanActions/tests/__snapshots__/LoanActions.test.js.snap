// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoanActions should render a loader when trigger an action 1`] = `
<LoanActions
  actionLoading={true}
  actions={Object {}}
  data={Object {}}
  onAction={[Function]}
>
  <List
    horizontal={true}
  >
    <div
      className="ui horizontal list"
      role="list"
    />
  </List>
</LoanActions>
`;

exports[`LoanActions should render an List.Item for each available action 1`] = `
<LoanActions
  actions={Object {}}
  availableActions={
    Object {
      "action": "url",
      "anotheraction": "another-url",
    }
  }
  data={
    Object {
      "title": "A title",
    }
  }
  onAction={[Function]}
>
  <List
    horizontal={true}
  >
    <div
      className="ui horizontal list"
      role="list"
    />
  </List>
</LoanActions>
`;

exports[`LoanActions should render initial state and check props 1`] = `
<LoanActions
  actions={Object {}}
  data={Object {}}
  onAction={[Function]}
>
  <List
    horizontal={true}
  >
    <div
      className="ui horizontal list"
      role="list"
    />
  </List>
</LoanActions>
`;
